contrib/bloop/src/mill.contrib.bloop/BloopImpl.scala

```diff
@@ -0,0 +1,339 @@
+package mill.contrib.bloop
+
+import ammonite.ops._
+import bloop.config.ConfigEncoderDecoders._
+import bloop.config.{Config => BloopConfig}
+import mill._
+import mill.api.Loose
+import mill.define.{Module => MillModule, _}
+import mill.eval.Evaluator
+import mill.scalalib._
+import os.pwd
+
+/**
+  * Implementation of the Bloop related tasks. Inherited by the
+  * `mill.contrib.Bloop` object, and usable in tests by passing
+  * a custom evaluator.
+  */
+class BloopImpl(ev: () => Evaluator, wd: Path) extends ExternalModule {
+
+  /**
+    * Generates bloop configuration files reflecting the build,
+    * under pwd/.bloop.
+    */
+  def install = T {
+    Task.traverse(computeModules)(writeBloopConfig)
+  }
+
+  /**
+    * Trait that can be mixed-in to quickly access the bloop config
+    * of the module.
+    *
+    * {{{
+    * object myModule extends ScalaModule with Bloop.Module {
+    *    ...
+    * }
+    * }}}
+    */
+  trait Module extends MillModule with CirceCompat { self: JavaModule =>
+
+    object bloop extends MillModule {
+
+      def config = T {
+        // Forcing the generation of the config for all modules here
+        val installed: Seq[(String, Path)] = install()
```

Comment:
This call the reason for the path-dependant trait. Refencing the global `mill.contrib.Bloop` module would disallow overrides and make testing fairly impossible.